reference/debugging.md

```diff
@@ -0,0 +1,229 @@
+        4 -rw-r--r--    1 root     root          3322 Oct 15 12:26 envoy-1.json
+        8 -rw-r--r--    1 root     root          4101 Oct 15 12:35 envoy-2.json
+        8 -rw-rw-r--    1 root     root          5245 Sep 25 20:28 hot-restarter.py
+    20 -rw-rw-r--    1 root     root         16952 Sep 25 20:28 kubewatch.py
+        4 -rwxrwxr--    1 root     root           175 Sep 25 20:28 requirements.txt
+        4 -rwxr-xr-x    1 root     root           997 Sep 25 20:28 start-envoy.sh
+    ```
+
+The `ambassador-config-N` lines indicate the directories which contain the specific Ambassador configuration used in YAML files during updates. Higher numbers indicate the most recent configuration, as verified by the directory timestamp.
+
+2. Determine the latest configuration. If you haven't overridden the configuration directory, the latest configuration will be in `/ambassador/snapshots`. If you have overridden it, Ambassador saves  configurations in `$AMBASSADOR_CONFIG_BASE_DIR/snapshots`.
+
+    In the snapshots directory:
+
+    * `snapshot.yaml` contains the full input configuration that Ambassador has found;
+    * `aconf.json` contains the Ambassador configuration extracted from the snapshot;
+    * `ir.json` contains the IR constructed from the Ambassador configuration; and
+    * `econf.json`contains the Envoy configuration generated from the IR.
+
+    The Envoy configuration is then split into `$AMBASSADOR_CONFIG_BASE_DIR/bootstrap-ads.json` and `$AMBASSADOR_CONFIG_BASE_DIR/envoy/envoy.json`, which are the actual input files handed to Envoy.
+
+    In the snapshots directory, the current configuration will be stored in files with no digit suffix, and older configurations have increasing numbers. For example, `ir.json` is current, `ir-1.json` is the next oldest, then `ir-2.json`, etc.
+
+5. If something is wrong with `snapshot` or `aconf`, there is an issue with your configuration. If something is wrong with `ir` or `econf`, you should [open an issue on Github](https://github.com/datawire/ambassador/issues/new/choose).
+6. Next, find the corresponding Envoy file. The file will be titled `envoy-N.json` where N matches the number of the `ambassador-config-N` directory number.
+7. Print the contents of the corresponding Envoy file that is generated during Ambassador configuration with the following command: `/ambassador # cat envoy-2.json`
+
+    The terminal will print something similar to the following:
+
+    ```console
+    /ambassador # cat envoy-2.json
+
+    {
+    "listeners": [
+
+    {
+        "address": "tcp://0.0.0.0:8080",
+
+        "filters": [
+        {
+            "type": "read",
+            "name": "http_connection_manager",
+            "config": {"codec_type": "auto",
+            "stat_prefix": "ingress_http",
+            "access_log": [
+                {
+    ```
+
+The contents of the Envoy configuration files can be very useful when looking for subtle mapping issues or bugs.
```

Comment:
Reading the Envoy config can still be useful, but they should look in `$AMBASSADOR_CONFIG_BASE_DIR/envoy/envoy.json` for the main config, and `$AMBASSADOR_CONFIG_BASE_DIR/bootstrap-ads.json` for the bootstrap config (which is important only for things stats, logging, and auth).